teuthology/provision/maas.py

```diff
@@ -0,0 +1,345 @@
+
+# Set default values
+TIMEOUT = 900 # 15 minutes
+
+
+def enabled(warn=False):
+    """Check for required MaaS settings
+
+    :param warn: Whether or not to log a message containing unset parameters
+
+    :returns: True if they are present; False if they are not
+    """
+    maas_conf = config.get("maas", dict())
+    params = ["endpoint", "api_key", "machine_types"]
+    unset = [param for param in params if not maas_conf.get(param)]
+    if unset and warn:
+        log.warning(
+            "MaaS disabled; set the following config options to enable: %s",
+            " ".join(unset),
+        )
+
+    if unset:
+        api_key = config.get("maas", dict()).get("api_key", "")
+        if api_key and api_key.split(":") < 3:
+            log.warning(
+                "MaaS api_key appears to be malformed; expected format is "
+                "'consumer_key:consumer_token:secret'"
+            )
+            return False
+
+    return True
+
+
+def get_types():
+    """Fetch and parse maas machine_types config
+
+    :returns: The list of MaaS-configured machine types. 
+              An empty list if MAAS is not configured.
+    """
+    if not enabled():
+        return []
+    maas_conf = config.get("maas", dict())
+    types = maas_conf.get("machine_types", "")
+    if not isinstance(types, list):
+        types = types.split(",")
+    return [type_ for type_ in types if type_]
+
+def release_machine(name):
+    """Release a machine from MAAS management
+
+    :param name: The fully-qualified domain name of the machine to release
+    """
+    MAAS(name, os_type="", os_version="").release()
+
+
+class MAAS(object):
+    """Reimage machines with https://maas.io"""
+
+    def __init__(self, name,  os_type, os_version):
+        """Initialize the MAAS object
+
+        :param name: The fully-qualified domain name of the machine to manage
+        :param os_type: The OS type to deploy (e.g. "ubuntu")
+        :param os_version: The OS version to deploy (e.g. "noble")
+        """
+        if not enabled():
+            raise RuntimeError("MAAS is not configured!")
+
+        self.remote = teuthology.orchestra.remote.Remote(
+            misc.canonicalize_hostname(name))
+        self.name = self.remote.hostname
+        self.shortname = self.remote.shortname
+        self.os_type = os_type
+        self.os_version = os_version
+
+        self.log = log.getChild(self.shortname)
+
+        self.session = self._session()
+        self.system_id = self.get_host_data()["system_id"]
+
+    def _session(self):
+        """Create an OAuth1Session for communicating with the MaaS server"""
+        key, token, secret = config.maas["api_key"].split(":")
+        return OAuth1Session(
+            key,
+            resource_owner_key=token,
+            resource_owner_secret=secret,
+            signature_method=SIGNATURE_PLAINTEXT
+        )
+
+    def do_request(
+            self, url_suffix, method="GET", params={}, data={}, verify=True
```

Comment:
Watch out for mutable defaults as they have surprising consequences: https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

Reply:
Thanks @zmc ,

Used `None` instead of empty dict `{}`.